src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs

```diff
@@ -1119,6 +1119,9 @@ private Type GetCompatibleGenericBaseClass(Type type, Type baseType)
                 }
             }
 
+            private static bool PropertyIsConstructorParameter(PropertyGenerationSpec propSpec, ParameterGenerationSpec[]? paramGenSpecArray)
+                => paramGenSpecArray != null && paramGenSpecArray.Any(paramSpec => propSpec.ClrName == paramSpec.ParameterInfo.Name);
```

Comment:
Shouldn't this also check that the types are matching?

Reply:
Hmm no I wouldn't think so.

If we have a `record`, it would already be a compile-time error to give the property a different type than the constructor parameter.

Of course that's not true at all if we have a regular `class`, such as:

```
public class MyClass
{
    public MyClass(int Value)
    {
        this.Value = Value.ToString();
    }

    public string Value { get; init; }
}
```

But how will STJ handle this scenario? The deserializer **has** to pass the value in as a constructor parameter - it has no choice, since that's the constructor it has decided to call. It's not going to use the property setter, so it doesn't matter what type it has.

Reply:
The deserializer would fail at runtime complaining about mismatch between property and parameter. Technically, this means the property is not associated with the constructor and is part of the list of settable properties in the contract, so emitting a warning would be accurate. But it probably doesn't matter much in practice.

Reply:
> The deserializer would fail at runtime complaining about mismatch between property and parameter.

Eek! That sounds like it deserves its own diagnostic and warning message to communicate the type mismatch problem, since we're saying that deserialization would fail even if the property setter weren't init only.

Reply:
That's definitely something we could implement in the future. Related to https://github.com/dotnet/runtime/issues/44428